- Roles, Refresh Tokens, Sign out

John-Emad opened the pull request:
- Adding/Removing Admin, Owner roles
- Refresh Token implementation
- Signout users

John-Emad committed:
- Adding/Removing Admin, Owner roles
- Refresh Token implementation
- Signout users

John-Emad committed:
Merge branch 'dev' into AddRoles-AddRefreshToken

omar98alaa commented on Eskon.API/Controllers/UserController.cs:
```diff
@@ -41,6 +43,42 @@ public async Task<IActionResult> GetStudentById([FromRoute] Guid id)
             return NewResult(response);
         }
 
+        [Authorize(Roles = "Customer, Admin")]
```
No need for specific authorization as every user is a Customer and anyone can become an Owner.
Make it so that the customer assigns himself as Owner... No need for admin assignment.
However, you may add some logic to skip processing if user is already an Owner but it is not necessary.

John-Emad replied:
What about the Root role? for that reason a specific authorization is written, if the root user will be authorized to that end point, then there is no need for specific authorizations as you said..

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Handler/UserCommandHandler.cs:
```diff
@@ -1,41 +1,42 @@
+        , IRequestHandler<AddOwnerRoleToUserCommand, Response<TokenResponseDto>>
+        , IRequestHandler<DeleteOwnerRoleFromUserCommand, Response<TokenResponseDto>>
+        , IRequestHandler<AddAdminRoleToUserCommand, Response<TokenResponseDto>>
+        , IRequestHandler<DeleteAdminRoleFromUserCommand, Response<TokenResponseDto>>
```
Would be better if you make an `IUserCommandHandler` interface that just extends all these `IRequestHandler` interfaces so that the code will be more readable.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Handler/UserCommandHandler.cs:
```diff
@@ -1,41 +1,42 @@
 
         #region Constructors
-        public UserCommandHandler(IUserService userService, IMapper mapper, IAuthenticationService authenticationService, UserManager<User> userManager, RoleManager<IdentityRole<Guid>> roleManager, SignInManager<User> signInManager)
+        public UserCommandHandler(IUserService userService, IMapper mapper, IAuthenticationService authenticationService, UserManager<User> userManager, RoleManager<IdentityRole<Guid>> roleManager, SignInManager<User> signInManager, IRefreshTokenService refreshTokenService)
```
Is it time to consider some sort of unit of work ? :)

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Command/AddAdminRoleToUserCommand.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Eskon.Core.Features.UserFeatures.Commands.Command
+{
+    public record AddAdminRoleToUserCommand(Guid UserToBeAdminId) : IRequest<Response<TokenResponseDto>>;
```
Adding role to user should not return a token. It will be returned to the root user not the new admin which is not what intended.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Command/DeleteAdminRoleFromUserCommand.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace Eskon.Core.Features.UserFeatures.Commands.Command
+{
+    public record DeleteAdminRoleFromUserCommand(Guid UserToRemoveAdminId) : IRequest<Response<TokenResponseDto>>;
```
No need for TokenResponse here as well.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Command/DeleteOwnerRoleFromUserCommand.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace Eskon.Core.Features.UserFeatures.Commands.Command
+{
+    public record DeleteOwnerRoleFromUserCommand(Guid UserToRemoveOwnerId) : IRequest<Response<TokenResponseDto>>;
```
No need for TokenResponse here as well.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Handler/UserCommandHandler.cs:
```diff
@@ -69,25 +70,197 @@ public UserCommandHandler(IUserService userService, IMapper mapper, IAuthenticat
+            var oldToken = await _refreshTokenService.GetTokenByUserIdAsync(user.Id);
+            if (oldToken != null)
+            {
+                oldToken.IsRevoked = true;
```
Just delete the expired/signed out token from DB.

Optimization:
Make an In-memory storage for refresh tokens not in DB... so that if the server restarts (in case of an update for example) everyone should re-login.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Handler/UserCommandHandler.cs:
```diff
@@ -69,25 +70,197 @@ public UserCommandHandler(IUserService userService, IMapper mapper, IAuthenticat
+            var user = await _userManager.FindByIdAsync(request.UserToBeOwnerId.ToString());
+
+            if (user == null)
+                return BadRequest<TokenResponseDto>("User Not Found");
```
Use `NotFound` response instead.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Handler/UserCommandHandler.cs:
```diff
@@ -69,25 +70,197 @@ public UserCommandHandler(IUserService userService, IMapper mapper, IAuthenticat
+
+            var existingRefreshToken = await _refreshTokenService.GetTokenByUserIdAsync(user.Id);
+
+            if (existingRefreshToken == null || existingRefreshToken.ExpiresAt < DateTime.UtcNow)
```
A user reaching this endpoint should be already authenticated. Just return the new access token with the updated role.

John-Emad replied:
I think that it is a type of security the User is already authenticated surely but we don't know from when he is logged, what if it's token had expired? so there is an added validation..

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Handler/UserCommandHandler.cs:
```diff
@@ -69,25 +70,197 @@ public UserCommandHandler(IUserService userService, IMapper mapper, IAuthenticat
+        #endregion
+
+        #region Delete Owner Role
+        public async Task<Response<TokenResponseDto>> Handle(DeleteOwnerRoleFromUserCommand request, CancellationToken cancellationToken)
```
No need to return tokens here.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Handler/UserCommandHandler.cs:
```diff
@@ -69,25 +70,197 @@ public UserCommandHandler(IUserService userService, IMapper mapper, IAuthenticat
+        #endregion
+
+        #region Add Admin Role
+        public async Task<Response<TokenResponseDto>> Handle(AddAdminRoleToUserCommand request, CancellationToken cancellationToken)
```
No need to return tokens here too.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Handler/UserCommandHandler.cs:
```diff
@@ -69,25 +70,197 @@ public UserCommandHandler(IUserService userService, IMapper mapper, IAuthenticat
-            return Success(accessToken);
-        } 
+        #region Delete Admin Role
+        public async Task<Response<TokenResponseDto>> Handle(DeleteAdminRoleFromUserCommand request, CancellationToken cancellationToken)
```
No need to return tokens.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Utilities/GetRefreshTokenHandler.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace Eskon.Core.Features.UserFeatures.Utilities
+{
+    public class GetRefreshTokenHandler :ResponseHandler, IRequestHandler<GetRefreshToken, Response<TokenResponseDto>>
```
Rename `GetRefreshToken` to something like `GetNewAccessToken` as the current name is confusnig.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Utilities/GetRefreshTokenHandler.cs:
```diff
@@ -0,0 +1,46 @@
+        {
+            var storedToken = await _refreshTokenService.GetStoredTokenAsync(request.token);
+
+            if (storedToken == null || storedToken.IsRevoked || storedToken.ExpiresAt < DateTime.UtcNow)
```
No need to check for revoked as invalid refresh tokens should be deleted.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Utilities/GetRefreshTokenHandler.cs:
```diff
@@ -0,0 +1,46 @@
+
+            if (storedToken == null || storedToken.IsRevoked || storedToken.ExpiresAt < DateTime.UtcNow)
+            {
+                return BadRequest<TokenResponseDto>("Refresh token is invalid or expired. Please login again.");
```
Return `401 Unauthorized` so it would be easier to detect in frontend.

omar98alaa commented on Eskon.Core/Features/UserFeatures/Commands/Handler/UserCommandHandler.cs:
```diff
@@ -69,25 +70,197 @@ public UserCommandHandler(IUserService userService, IMapper mapper, IAuthenticat
+                oldToken.IsRevoked = true;
+                await _refreshTokenService.SaveChangesAsync();
+            }
+            var accessToken = await _authenticationService.GenerateJWTTokenAsync(user);
```
Optimization:
Make `GenerateJWTTokenAsync` take the user role (or any indicator) as an argument, and modify the implementation so that the expiry time for an admin is not the same for other users.

John-Emad replied:
Will solve it at the GenerateJWTTokenAsync() method itself

omar98alaa commented on Eskon.Domian/Entities/Identity/User.cs:
```diff
@@ -21,6 +22,10 @@ public class User : IdentityUser<Guid>, IBaseModel
 
+        public string? Code { get; set; }
+        [InverseProperty(nameof(UserRefreshToken.User))]
+        public virtual ICollection<UserRefreshToken> UserRefreshTokens { get; set; } = new HashSet<UserRefreshToken>();
```
Would need to be changed if token will no longer be stored in DB.

John-Emad replied:
Deleting Refresh Tokens (Hard Delete)
will Keep the token table small, and no special handling as will deal only with valid existing tokens but, there will be 
No Historical Record as we lose track of which tokens were revoked, when, or why. Moreover, 
Potential Security Risk Issues as a Reuse of deleted tokens might go unnoticed, especially if deletion logic fails or is bypassed.
So, in my opinion keeping it for future work as logging and security tracking would be plus

omar98alaa commented on Eskon.Service/Services/AuthenticationService.cs:
```diff
@@ -14,58 +15,59 @@ public class AuthenticationService : IAuthenticationService
                 _jwtSettings.Audience,
                 claims,
-                expires: DateTime.Now.AddDays(2),
+                expires: DateTime.Now.AddMinutes(8),
```
Make this different for admin.

omar98alaa reviewed (changes requested):
Great work overall.
Refresh token logic needs some tweaking.

John-Emad committed:
- Separating User Roles handler, User Account handler into two handlers.
- Managing Admin Access token Expiry date.

John-Emad committed:
Merge branch 'AddRoles-AddRefreshToken' of https://github.com/omar98alaa/Eskon into AddRoles-AddRefreshToken

John-Emad committed:
- Editing the Controllers withe the new namings of the handlers and namespaces.
- Renaming Wrong Spelled Reposotry to Repository at ReviewRepository classes (Service, Repository) and interfaces

omar98alaa commented on Eskon.Core/Features/AccountFeatures/Commands/Handler/UserAccountCommandHandler.cs:
```diff
@@ -0,0 +1,122 @@
+            if (stored != null)
+            {
+                stored.IsRevoked = true;
+                await _refreshTokenService.SaveChangesAsync();
```
Revoke service here as well.

omar98alaa commented on Eskon.Core/Features/UserRolesFeatures/Commands/Handler/UserRolesCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+            var user = await _userManager.FindByIdAsync(request.UserToBeOwnerId.ToString());
+
+            if (user == null)
+                return BadRequest<TokenResponseDto>("User Not Found");
```
Return `NotFound` instead.

omar98alaa commented on Eskon.Core/Features/UserRolesFeatures/Commands/Handler/UserRolesCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+
+            var existingRefreshToken = await _refreshTokenService.GetTokenByUserIdAsync(user.Id);
+
+            if (existingRefreshToken == null || existingRefreshToken.ExpiresAt < DateTime.UtcNow)
```
I understand your concern about security measures but my point is that validating the refresh token here will violate seperation of concerns. Refresh tokens are validated when used to get new access token @ /Account/Refresh.
A user reaching this endpoint means he has a valid access token which is enough.

omar98alaa commented on Eskon.Core/Features/UserRolesFeatures/Commands/Handler/UserRolesCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+            var user = await _userManager.FindByIdAsync(request.UserToRemoveOwnerId.ToString());
+
+            if (user == null)
+                return BadRequest<TokenResponseDto>("User Not Found");
```
NotFound response.

omar98alaa commented on Eskon.Core/Features/UserRolesFeatures/Commands/Handler/UserRolesCommandHandler.cs:
```diff
@@ -0,0 +1,147 @@
+        #endregion
+
+        #region Delete Owner Role
+        public async Task<Response<TokenResponseDto>> Handle(DeleteOwnerRoleFromUserCommand request, CancellationToken cancellationToken)
```
No need to generate tokens here as the user is not deleting his own role.

omar98alaa commented on Eskon.Service/Services/AuthenticationService.cs:
```diff
@@ -14,58 +15,60 @@ public class AuthenticationService : IAuthenticationService
         #region Fields
         private readonly JwtSettings _jwtSettings;
         private readonly UserManager<User> _userManager;
```
Authentication service should not depend on UserManager

John-Emad committed:
- Removing deleteRole end point
- Editing Authentication service
- Editing Addowner end point to take the Id from the token
- GetNewAccessToken instead of GetRefreshToken

omar98alaa commented on Eskon.Service/Services/AuthenticationService.cs:
```diff
@@ -14,61 +14,58 @@ public class AuthenticationService : IAuthenticationService
-            var claims = await GetClaims(user);
+            var userAllClaims = GeneratedAllUserClaims(user, userManagerRoles, userManagerClaims);
+            bool isAdmin = false;
+            foreach (var role in userManagerRoles)
```
You can use `bool isAdmin = userManagerRoles.Contains("Admin");`

omar98alaa commented on Eskon.Service/Services/AuthenticationService.cs:
```diff
@@ -14,61 +14,58 @@ public class AuthenticationService : IAuthenticationService
                 new Claim(ClaimTypes.Name, user.UserName)
             };
-            foreach (var role in roles)
+            foreach (var role in userManagerRoles)
```
You can use `claims.AddRange(userManagerRoles.Select(role => new Claim(ClaimTypes.Role, role));`

John-Emad replied:
cannot be done due to different conversion types

omar98alaa reviewed (changes requested):
Great work John. Just a minor adjustment.

John-Emad committed:
using .Contains instead of looping items at userManagerRoles

omar98alaa reviewed (approved):
Well done John. Excellent work.

omar98alaa committed:
Merge branch 'dev' into AddRoles-AddRefreshToken

omar98alaa committed:
Merge branch 'dev' into AddRoles-AddRefreshToken
